hive/agents/ppo.py

```diff
@@ -0,0 +1,405 @@
+
+        if self.rew_norm:
+            self.rew_norm.update(update_info["reward"], update_info["done"])
+            update_info["reward"] = self.rew_norm(update_info["reward"])
+
+        preprocessed_update_info = {
+            "observation": update_info["observation"],
+            "action": update_info["action"],
+            "reward": update_info["reward"],
+            "done": update_info["done"],
+            "logprob": self._act_vars["logprob"],
+            "values": self._act_vars["value"],
+            "returns": np.empty(self._act_vars["value"].shape),
+            "advantages": np.empty(self._act_vars["value"].shape),
+        }
+        if "agent_id" in update_info:
+            preprocessed_update_info["agent_id"] = int(update_info["agent_id"])
+
+        return preprocessed_update_info
+
+    def preprocess_update_batch(self, batch):
+        """Preprocess the batch sampled from the replay buffer.
+
+        Args:
+            batch: Batch sampled from the replay buffer for the current update.
+
+        Returns:
+                - Preprocessed batch.
+        """
+        for key in batch:
+            batch[key] = torch.tensor(batch[key], device=self._device)
+
+        return batch
+
+    @torch.no_grad()
+    def get_action_logprob_value(self, observation):
+        """Returns the action, logprob, and value for the agent
+
+        Args:
+            observation: The current observation.
+        """
+        observation = torch.tensor(
+            np.expand_dims(observation, axis=0), device=self._device
+        ).float()
+        action, logprob, _, value = self._actor_critic(observation)
+        action = action.cpu().detach().numpy()
+        logprob = logprob.cpu().numpy()
+        value = value.cpu().numpy()
+        action = action[0]
+
+        return action, logprob, value
+
+    @torch.no_grad()
+    def act(self, observation):
+        """Returns the action for the agent.
+
+        Args:
+            observation: The current observation.
+        """
+        if self.obs_norm:
+            self.obs_norm.update(observation)
+            observation = self.obs_norm(observation)
+        action, logprob, value = self.get_action_logprob_value(observation)
+        self._act_vars = dict(logprob=logprob, value=value)
+        return action
+
+    def update(self, update_info):
+        """
+        Updates the PPO agent.
+
+        Args:
+            update_info: dictionary containing all the necessary information to
+                update the agent. Should contain a full transition, with keys for
+                "observation", "action", "reward", and "done".
+        """
+        if not self._training:
+            return
+
+        # Add the most recent transition to the replay buffer.
+        self._replay_buffer.add(**self.preprocess_update_info(update_info))
+
+        if self._replay_buffer.size() + 1 == self._transitions_per_update:
```

Comment:
I'd maybe write this logic as 
```suggestion
        if self._replay_buffer.size() == self._transitions_per_update - 1:
```